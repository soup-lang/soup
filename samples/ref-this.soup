func hex2(n)
    s: (n & 255).toString(16)
    if s.length = 1
        return '0' + s
    return s

func Color(r, g, b)
    this.r: Math.floor(r)
    this.g: Math.floor(g)
    this.b: Math.floor(b)
    this["to-string"]: (): '#' + hex2(this.r) + hex2(this.g) + hex2(this.b)
    this.toString: this['to-string']

red: *Color(255, 0, 0)
console.log(red.toString())

teal: *Color(0, 255, 255)
console.log(teal['to-string']())

whatTheHell: *Color(345, 567, 10.24)
console.log(whatTheHell["to-string"]())

func logLines(lines)
    lines |:
        this.line: $
        setTimeout(%, 1)
        console.log(this.line)
    console.log(this.line)
    console.log('Done')

logLines(['a', 'b', 'c'])

console.log('End')
