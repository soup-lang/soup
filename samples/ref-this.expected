#ff0000
#00ffff
#59370a
End
a
b
c
c
Done
